Merchants can change their deposit schedule

brucealdridge opened the pull request:
Fixes #4464

#### Changes proposed in this Pull Request

With this PR, merchants will be able to change their deposit schedule between Daily, Weekly and, Monthly. When opting for a weekly schedule a desired day of the week can be chosen, with a monthly schedule a day of the month can be chosen.

A new `Deposits` section is added to WCPay Settings where the frequency can be selected.

![image](https://user-images.githubusercontent.com/57298/187828238-b057cdbf-a08a-42af-9cc9-9954c9ee3cd6.png)

![image](https://user-images.githubusercontent.com/57298/187828269-8cd678f1-86d2-44d3-83f5-ffb3ea4accb6.png)

![image](https://user-images.githubusercontent.com/57298/187828295-0e247afd-31c3-4420-862b-c53957a7d87b.png)

If an account is new or deposit scheduling is unavailable we show a warning.

![image](https://user-images.githubusercontent.com/57298/187828559-4a906317-41b1-419c-9184-224905e82243.png)

![image](https://user-images.githubusercontent.com/57298/187828714-671b007c-4c7b-4ed1-96e6-a5e7b43e88e9.png)

#### Testing instructions

Testing of deposits is currently limited to live accounts only. Test accounts will see a warning message. 
Alternatively, with a development version of the Server, testing is available.

1. Visit WC Pay Settings.
2. Change the deposit schedule.
3. Verify the deposit schedule in stripe dashboard.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

brucealdridge committed:
Merchants can change their deposit schedule

Commit:
Merge branch 'develop' into add/4464-merchants-can-change-deposit-schedule

aprea commented on tests/unit/admin/test-class-wc-rest-payments-settings-controller.php:
```diff
@@ -454,6 +457,10 @@ public function test_update_settings_saves_account() {
 		$request->set_param( 'account_branding_icon', 'test branding_icon' );
 		$request->set_param( 'account_branding_primary_color', 'test branding_primary_color' );
 		$request->set_param( 'account_branding_secondary_color', 'test branding_secondary_color' );
+		$request->set_param( 'account_branding_primary_color', 'test branding_primary_color' );
```
This line is copypasta

aprea commented on includes/class-wc-payments-account.php:
```diff
@@ -302,6 +302,66 @@ public function get_branding_secondary_color() : string {
+	 *
+	 * @return string  e.g. 2, 7.
+	 */
+	public function get_deposit_delay_days(): string {
```
Should the return type be `int`?

aprea commented on includes/class-wc-payments-account.php:
```diff
@@ -302,6 +302,66 @@ public function get_branding_secondary_color() : string {
+	 */
+	public function get_deposit_delay_days(): string {
+		$account = $this->get_cached_account_data();
+		return $account['deposits']['delay_days'] ?? '';
```
Should we use a fallback value of `7` rather than an empty string?

brucealdridge replied:
I've updated this to return an `int|null`. Just to limit the use of magic numbers.
This means we're using the `7` magic number in two places. I haven't checked, but we should be able to remove it from the javascript.
https://github.com/Automattic/woocommerce-payments/blob/52f17d2abf40c700c2885ba4cf55bd67af8a8b2e/includes/class-wc-payment-gateway-wcpay.php#L2262
and 
https://github.com/Automattic/woocommerce-payments/blob/52f17d2abf40c700c2885ba4cf55bd67af8a8b2e/client/data/settings/selectors.js#L107

aprea commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -2181,6 +2198,114 @@ protected function get_account_branding_secondary_color( $default_value = '' ):
+	 *
+	 * @return string number of days.
+	 */
+	protected function get_deposit_delay_days( string $empty_value = '' ): string {
```
Similar feedback as per below. Should we use `int` for the return type and `7` as the empty value? I assume the API will return this value as an integer rather than a string?

Apologies if this assumption is incorrect!

aprea commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -2181,6 +2198,114 @@ protected function get_account_branding_secondary_color( $default_value = '' ):
+				return $this->account->get_deposit_delay_days();
+			}
+		} catch ( Exception $e ) {
+			Logger::error( 'Failed to get deposit schedule monthly anchor.' . $e );
```
Copypasta, it should probably be:
>Failed to get deposit delay days.

(or similar)

aprea commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -2181,6 +2198,114 @@ protected function get_account_branding_secondary_color( $default_value = '' ):
+	}
+
+	/**
+	 * Gets connected account deposit status.
```
Copypasta here, in the return description and in the error log. It should reference the waiting period rather than the deposit status.

aprea commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -2181,6 +2198,114 @@ protected function get_account_branding_secondary_color( $default_value = '' ):
+	 *
+	 * @return string deposit status or default value.
+	 */
+	protected function get_deposit_completed_waiting_period( string $empty_value = '' ): bool {
```
I think the empty value will need to be a boolean as the function uses `bool` as the return type.

james-allan commented on includes/class-wc-payments-account.php:
```diff
@@ -302,6 +302,66 @@ public function get_branding_secondary_color() : string {
+	}
+
+	/**
+	 * Gets the whether the account has completed waiting period
```
```suggestion
	 * Gets whether the account has completed the deposit waiting period.
```

james-allan commented on includes/class-wc-payments-account.php:
```diff
@@ -302,6 +302,66 @@ public function get_branding_secondary_color() : string {
+	/**
+	 * Gets the deposit status
+	 *
+	 * @return string  e.g. disblaed, blocked, enabled.
```
Small typo
```suggestion
	 * @return string  e.g. disabled, blocked, enabled.
```

aprea commented on includes/admin/class-wc-rest-payments-settings-controller.php:
```diff
@@ -363,6 +379,12 @@ public function get_settings(): WP_REST_Response {
 				'platform_checkout_store_logo'      => $this->wcpay_gateway->get_option( 'platform_checkout_store_logo' ),
+				'deposit_schedule_interval'         => $this->wcpay_gateway->get_option( 'deposit_schedule_interval' ),
+				'deposit_monthly_anchor'            => $this->wcpay_gateway->get_option( 'deposit_schedule_monthly_anchor' ),
+				'deposit_weekly_anchor'             => $this->wcpay_gateway->get_option( 'deposit_schedule_weekhly_anchor' ),
```
Typo: `deposit_schedule_weekhly_anchor`

```suggestion
				'deposit_weekly_anchor'             => $this->wcpay_gateway->get_option( 'deposit_schedule_weekly_anchor' ),
```

aprea commented on includes/admin/class-wc-rest-payments-settings-controller.php:
```diff
@@ -157,6 +161,18 @@ public function register_routes() {
+					],
+					'deposit_schedule_monthly_anchor'   => [
+						'description' => __( 'Monthly anchor for deposit scheduling when interval is set to monthly', 'woocommerce-payments' ),
+						'type'        => 'string',
```
Should the type be `integer`? I believe we're sending the calendar date here, e.g. `1`, `2`, `3` ... `31`, etc.

brucealdridge replied:
I've left as `string` as empty values are sent through as an empty string rather than null which triggers an error.

brucealdridge committed:
Style changes

aprea commented on client/settings/deposits/index.js:
```diff
@@ -0,0 +1,217 @@
+		setDepositScheduleMonthlyAnchor,
+	] = useDepositScheduleMonthlyAnchor();
+
+	const daysOfWeek = [
```
**Nitpick:** `daysOfWeek` and `monthlyAnchors` can be defined outside of the component as they do not use any props or state.

Commit:
Merge branch 'develop' into add/4464-merchants-can-change-deposit-schedule

Commit:
Update the deposits schedule controls heading to "Deposit schedule".

Commit:
The weekly anchor must be a business day.

Commit:
Update the "Last day of the month" option label to match the designs

Commit:
Move constants outside of the component.

Commit:
Fix 'weekhly' typo.

Commit:
Update the weekly / monthly REST payload keys to match the client expectations.

Commit:
Remove copypasta.

Commit:
Fix 'disblaed' typo.

Co-authored-by: James Allan <james.allan@automattic.com>

Commit:
Update function description for better readability.

Co-authored-by: James Allan <james.allan@automattic.com>

Commit:
Merge branch 'develop' into add/4464-merchants-can-change-deposit-schedule

Commit:
Add changelog entry.

Commit:
Merge branch 'add/4464-merchants-can-change-deposit-schedule' of https://github.com/Automattic/woocommerce-payments into add/4464-merchants-can-change-deposit-schedule

Commit:
Update the get_deposit_completed_waiting_period() function docblock and types.

brucealdridge committed:
Changes from PR review

brucealdridge committed:
Ensure the deposit schedule interval is sent through to the API when the anchor is updated.

Commit:
Merge branch 'develop' into add/4464-merchants-can-change-deposit-schedule

Commit:
Reset weekly/monthly anchors with a handleIntervalChange handler

Jinksi commented on client/settings/deposits/index.js:
```diff
@@ -0,0 +1,230 @@
+			<p className="help-text">
+				{ 'monthly' === depositScheduleInterval &&
+					__(
+						'Deposits scheduled on a weekend will be sent on the next business day.',
```
Since this is a 'string' variable and not a function, it might be more clear to rename this to `depositStatus` rather than `getDepositStatus`.

aprea replied:
Fixed in f174838.

Jinksi committed:
Add tests for Deposits component

brucealdridge committed:
Add unit tests

Jinksi committed:
Add Deposits component test for unknown deposit status

Jinksi committed:
Rename variable to clarify type

Jinksi committed:
Remove extraneous comment

Jinksi commented on client/settings/deposits/index.js:
```diff
@@ -0,0 +1,230 @@
+];
+
+const monthlyAnchors = [
+	{ label: __( '1st', 'woocommerce-payments' ), value: 1 },
```
This might be an over-optimisation (and late in the game), but we may be able to make this more succinct by adding the localized ordinal suffix on render using moment.js. E.g:

https://github.com/Automattic/woocommerce-payments/blob/93a28552fb83d225ffb6830f969bf1561bb18951/client/deposits/utils/index.js#L81-L84

Jinksi replied:
I'm leaving this alternative here but I **don't** think we should change it.

👎  It's not as easy to read
👍 there is less to translate
👍 day suffixed should be automatically localised (not tested)
❓ I feel like there must be a WP function that exists for this already

```js
function getMonthlyAnchors() {
	// Create an array of options for the monthly anchor select control.
	const monthlyAnchors = _.range( 28 ).map( ( val ) => ( {
		value: val + 1,
		label: moment()
			.date( val + 1 )
			.format( 'Do' ),
	} ) );
	// Add the last option for the last day of the month.
	monthlyAnchors.push( {
		value: 31,
		label: __( 'Last day of the month', 'woocommerce-payments' ),
	} );

	return monthlyAnchors;
}
```

Instead of:

https://github.com/Automattic/woocommerce-payments/blob/93a28552fb83d225ffb6830f969bf1561bb18951/client/settings/deposits/index.js#L38-L71

james-allan reviewed (dismissed):
Superduper nit: I notice the select boxes are cutting off the 'y' tail 

<img width="240" alt="Screen Shot 2022-09-01 at 2 59 13 pm" src="https://user-images.githubusercontent.com/8490476/187834753-8e09fe37-2039-4db9-9905-05c9b0046128.png">

If I add a `line-height: normal;` it fixes it. 

<img width="255" alt="Screen Shot 2022-09-01 at 3 00 13 pm" src="https://user-images.githubusercontent.com/8490476/187834879-83189446-ec36-4bc9-be97-657be88b7a87.png">

Notice the y in 'Weekly' is complete but cut off in 'Thursday'.

Jinksi commented:
Tested on mobile: iOS 15.6.1 Safari on an iPhone SE.

The only issue I see is the line-height issue mentioned above: https://github.com/Automattic/woocommerce-payments/pull/4656#pullrequestreview-1092771569

<details>
<summary>Screenshots</summary>

![IMG_2310](https://user-images.githubusercontent.com/3147296/187835563-0229e145-f462-456b-a790-171e24d22490.PNG)

![IMG_2313](https://user-images.githubusercontent.com/3147296/187835533-bf3ad929-4182-4f44-94cf-2041fa629625.PNG)

![IMG_2312](https://user-images.githubusercontent.com/3147296/187835551-838a7d03-ab8d-47e1-be00-ea3d49f272c4.PNG)
</details>

Jinksi commented:
> Superduper nit: I notice the select boxes are cutting off the 'y' tail

@james-allan I think this is originating from [@wordpress/components](https://www.npmjs.com/package/@wordpress/components). Mind you, it has probably already been resolved, but we're using an old version`11.1.5` (latest is `20.0.0`).

Update: it looks like this issue https://github.com/WordPress/gutenberg/issues/38009. I upgraded to `@wordpress/components@20.0.0` but the issue remains.

Jinksi reviewed (approved):
🙌  LGTM!

Using 2379-gh-Automattic/woocommerce-payments-server branch and a live account in dev mode. After saving each change it was visible in Stripe dashboard:
- [x] updated schedule to Daily
- [x] updated schedule to weekly on Monday
- [x] updated schedule to weekly on Friday
- [x] updated schedule to monthly on the 1st
- [x] updated schedule to monthly on the 15th
- [x] updated schedule to monthly on the 28th
- [x] updated schedule to monthly on the 'last day of the month'
    - This updates Stripe to the 31st, which I'm assuming translates to 'last day of the month' for Stripe.

Using production WCPay-server (to bypass `is_proxied_automattician()`) with a live account **not** in dev mode:
- [x] account within the new account waiting period message:
    `Your first deposit will be held for 7 days. Deposit scheduling will be available after this period.`
- [x] deposits soft-blocked (via mc blog RC) message:
    `Deposit scheduling is currently unavailable for your store.`
- [x] deposits hard-blocked (via mc blog RC) message:
    `Deposit scheduling is currently unavailable for your store.`

As per https://github.com/Automattic/woocommerce-payments/pull/4656#issuecomment-1233747714 I also tested the UI on an iPhone SE iOS 15.6.1 Safari ✅ 

I mentioned in https://github.com/Automattic/woocommerce-payments/pull/4656#issuecomment-1233752337 that the `line-height` issue looks like it is originating from `@wordpress/components`: https://github.com/WordPress/gutenberg/issues/38009. I upgraded to `@wordpress/components@20.0.0` but the issue remains. It is probably best to address that issue in that package. It doesn't need to block this PR IMO :shipit:

Commit:
Don't hardcode daily deposits on overview page when account has no deposit

When a store has completed the new waiting period but doesn't have a deposit yet, the
deposit schedule was hardcode to daily on the Payments Overview page.

Now that we allow merchants to set their own deposit schedule after completing the new waiting period,
this commit replaces the "daily" interval that is hardcoded to use the correct schedule

Commit:
Fix unit tests

aprea reviewed (approved):
LGTM, approving. I commented my testing results on the server PR here → 2379-gh-Automattic/woocommerce-payments-server#issuecomment-1233709615

shendy-a8c commented on client/settings/deposits/index.js:
```diff
@@ -0,0 +1,230 @@
+};
+const DepositsSchedule = () => {
+	const depositStatus = useDepositStatus();
+	const completedWaitingPeriod = useCompletedWaitingPeriod();
```
See my comment below about `Your first deposit will be held for %s days`.

```suggestion
	const completedWaitingPeriod = useCompletedWaitingPeriod();
	const depositDelayDays = useDepositDelayDays();
```

shendy-a8c commented on client/settings/deposits/index.js:
```diff
@@ -0,0 +1,230 @@
+		return (
+			<Notice status="warning" isDismissible={ false }>
+				<span>
+					{ __(
+						'Your first deposit will be held for 7 days. Deposit scheduling will be available after this period.',
+						'woocommerce-payments'
+					) }
```
From https://github.com/Automattic/woocommerce-payments/issues/4464:
```
The "7" above is derived from deposits_delay_days
```

```suggestion
					{ sprintf(
						__(
							'Your first deposit will be held for %s days. Deposit scheduling will be available after this period.',
							'woocommerce-payments'
						),
						depositDelayDays
					) }
```

brucealdridge replied:
I thought the same thing originally but was corrected. The first deposit is always held for 7 days as an anti-fraud measure.
The `delay_days` is the delay between when a payment is received and when it is available for deposits and can vary by country.

see p1661909116777449/1661907696.353099-slack-C03KT8NLL13

shendy-a8c reviewed (commented):
Great work, @brucealdridge! Please see my feedback below.

Commit:
Merge branch 'develop' into add/4464-merchants-can-change-deposit-schedule

Commit:
Add a helper link to the deposit notices.

Commit:
Merge branch 'add/4464-merchants-can-change-deposit-schedule' of https://github.com/Automattic/woocommerce-payments into add/4464-merchants-can-change-deposit-schedule

brucealdridge committed:
Setting Monthly anchor to an int

james-allan commented:
I tested the changes introduced in [fdbeeac](https://github.com/Automattic/woocommerce-payments/pull/4656/commits/fdbeeacb33a614305bb2f8f0c11705b3aa208c22) and they look good and work for me. 

<img width="694" alt="Screen Shot 2022-09-02 at 10 42 30 am" src="https://user-images.githubusercontent.com/8490476/188035391-4d4f0fd9-73cb-45b5-9afb-b8f3dd3f8f2e.png">

<img width="717" alt="Screen Shot 2022-09-02 at 10 43 18 am" src="https://user-images.githubusercontent.com/8490476/188035465-9de62126-b600-4c09-9328-d513b7bf7a39.png">

brucealdridge committed:
Fixing psalm issues

brucealdridge commented on psalm-baseline.xml:
```diff
@@ -12,6 +12,12 @@
     <UndefinedMethod occurrences="1">
       <code>$intent</code>
     </UndefinedMethod>
+    <InvalidNullableReturnType occurrences="1">
+        <code>string|array|int|bool</code>
+    </InvalidNullableReturnType>
+    <NullableReturnStatement occurrences="1">
+        <code>$this->get_deposit_schedule_monthly_anchor()</code>
+    </NullableReturnStatement>
```
Have added these to the psalm-baseline instead of adding `null` as a return type to `\WC_Payment_Gateway_WCPay::get_option`. 
Adding `null` causes multiple issues in downstream functions.

brucealdridge committed:
Merge branch 'develop' into add/4464-merchants-can-change-deposit-schedule
